Source/Extensions/Blazorise.TreeView/Internal/_TreeViewNode.razor.cs

```diff
@@ -0,0 +1,207 @@
+        builder.Append( "b-tree-view-node-collapsed", !Expanded );
+
+        base.BuildClasses( builder );
+    }
+
+    private async Task ToggleNode( TreeViewNodeState<TNode> nodeState, bool refresh = true )
+    {
+        nodeState.Expanded = !nodeState.Expanded;
+
+        if ( nodeState.Expanded )
+        {
+            if ( !ExpandedNodes.Contains( nodeState.Node ) )
+            {
+                ExpandedNodes.Add( nodeState.Node );
+            }
+
+            await ExpandedNodesChanged.InvokeAsync( ExpandedNodes );
+
+            if ( nodeState.HasChildren )
+            {
+                var childNodes = GetChildNodesAsync is not null
+                    ? await GetChildNodesAsync( nodeState.Node )
+                    : GetChildNodes( nodeState.Node );
+
+                if ( !nodeState.Children.Select( x => x.Node ).AreEqual( childNodes ) )
+                {
+                    nodeState.Children = new();
+
+                    await foreach ( var childNodeState in childNodes.ToNodeStates( HasChildNodesAsync, HasChildNodes, ExpandedNodes.Intersect( childNodes ?? Enumerable.Empty<TNode>() ).Any() ) )
+                    {
+                        nodeState.Children.Add( childNodeState );
+                    }
+                }
+            }
+        }
+        else
+        {
+            ExpandedNodes.Remove( nodeState.Node );
+            await ExpandedNodesChanged.InvokeAsync( ExpandedNodes );
+        }
+
+        if ( refresh )
+        {
+            DirtyClasses();
+
+            await InvokeAsync( StateHasChanged );
+        }
+    }
+
+    public async Task ExpandAll()
+    {
+        foreach ( var nodeState in NodeStates ?? Enumerable.Empty<TreeViewNodeState<TNode>>() )
+        {
+            if ( HasChildNodes( nodeState.Node ) )
+            {
+                if ( !nodeState.Expanded )
+                    await ToggleNode( nodeState, false );
+
+                ExecuteAfterRender( async () =>
+                {
+                    if ( nodeState.ViewRef is not null )
+                        await nodeState.ViewRef.ExpandAll();
+                } );
+            }
+        }
+
+        DirtyClasses();
+
+        await InvokeAsync( StateHasChanged );
+    }
+
+    public async Task CollapseAll()
+    {
+        foreach ( var nodeState in NodeStates ?? Enumerable.Empty<TreeViewNodeState<TNode>>() )
+        {
+            if ( HasChildNodes( nodeState.Node ) )
+            {
+                if ( nodeState.Expanded )
+                    await ToggleNode( nodeState, false );
+
+                ExecuteAfterRender( async () =>
+                {
+                    if ( nodeState.ViewRef is not null )
+                        await nodeState.ViewRef.CollapseAll();
+                } );
+            }
+        }
+
+        DirtyClasses();
+
+        await InvokeAsync( StateHasChanged );
+    }
+
+    private Action<TNode, NodeStyling> ResolveNodeStylingAction( Action<TNode, NodeStyling> action )
```

Comment:
Is this because you can't pass nulls around or something? Wouldn't it be better to pass a null and do nothing in the other side?

Reply:
You're right. This was done by the original TreeView creator, but it seems to not be necessary, so I have removed it.